server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/dlp/DlpDomainRules.java

```diff
@@ -0,0 +1,310 @@
+ ****************************************************************/
+
+package org.apache.james.transport.matchers.dlp;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+
+import javax.mail.Address;
+import javax.mail.BodyPart;
+import javax.mail.Message;
+import javax.mail.MessagingException;
+import javax.mail.Multipart;
+import javax.mail.internet.MimeMessage;
+
+import org.apache.james.core.MailAddress;
+import org.apache.james.dlp.api.DLPConfigurationItem;
+import org.apache.james.dlp.api.DLPConfigurationItem.Targets;
+import org.apache.james.javax.MultipartUtil;
+import org.apache.james.mime4j.util.MimeUtil;
+import org.apache.james.util.OptionalUtils;
+import org.apache.mailet.Mail;
+
+import com.github.fge.lambdas.Throwing;
+import com.github.fge.lambdas.predicates.ThrowingPredicate;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableCollection;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMultimap;
+
+public class DlpDomainRules {
+
+    @VisibleForTesting static DlpDomainRules matchNothing() {
+        return DlpDomainRules.of(new Rule(DLPConfigurationItem.Id.of("always false"), (mail) -> false));
+    }
+
+    @VisibleForTesting static DlpDomainRules matchAll() {
+        return DlpDomainRules.of(new Rule(DLPConfigurationItem.Id.of("always true"), (mail) -> true));
+    }
+
+    private static DlpDomainRules of(Rule rule) {
+        return new DlpDomainRules(ImmutableList.of(rule));
+    }
+
+    public static DlpDomainRulesBuilder builder() {
+        return new DlpDomainRulesBuilder();
+    }
+
+    static class Rule {
+
+        interface MatcherFunction extends ThrowingPredicate<Mail> { }
+
+
+        private static Stream<String> asStringStream(Address[] addresses) {
+            return Arrays.stream(addresses).map(Rule::asString);
+        }
+
+        private static String asString(Address address) {
+            return MimeUtil.unscrambleHeaderValue(address.toString());
+        }
+
+        private static class ContentMatcher implements Rule.MatcherFunction {
+
+            private final Pattern pattern;
+
+            private ContentMatcher(Pattern pattern) {
+                this.pattern = pattern;
+            }
+
+            @Override
+            public boolean doTest(Mail mail) throws MessagingException, IOException {
+                return matchSubject(pattern, mail) || matchBody(pattern, mail);
+            }
+
+            private boolean matchSubject(Pattern pattern, Mail mail) throws MessagingException {
+                return getMessageSubjects(mail).anyMatch(pattern.asPredicate());
+            }
+
+            private Stream<String> getMessageSubjects(Mail mail) throws MessagingException {
```

Comment:
Why do you return `Stream<String>` instead of `Optional<String>`? Inside the method, it takes only one or zero element, not a collection, don't really understand why.